lib/Partitioner/Partitioner.cpp

```diff
@@ -569,68 +581,390 @@ void Partitioner::doPartitioning(Function *F, NodeToFunctionMap &mapping) {
+        auto *tmp = save->getPlaceholder();
+        placeholders[input] = tmp;
+        N.setNthInput(inp, tmp);
+      }
+    }
+  }
+
+  DAG dag;
+  dag.root = std::move(DAGRoot);
+  dag.nodes = std::move(nodes);
+  partitions_.push_back(std::move(dag));
+
+  // Update links between nodes in the cloned functions. Add placeholders (and
+  // save nodes) where a link crosses a partition boundary.
+  for (auto *subF : mapping.getPartitions()) {
+    for (auto &N : subF->getNodes()) {
+      for (int inp = 0, e = N.getNumInputs(); inp < e; inp++) {
+        auto input = N.getNthInput(inp);
+        if (isa<Storage>(input.getNode()))
+          continue;
+        // Link this node to the clone of its input.
+        auto *clone = currToNew[input.getNode()];
+        N.setNthInput(inp, NodeValue(clone, input.getResNo()));
+      }
+    }
+  }
+
+  // For all DAGNode without parents, link them to the root DAG.
+  for (auto *subF : mapping.getPartitions()) {
+    if (funcDAG[subF]->parents.size() == 0) {
+      funcDAG[subF]->parents.push_back(root);
+      root->children.push_back(funcDAG[subF]);
+    }
+  }
+}
+
+FunctionToBackendKindMapTy
+Partitioner::backendBasedPartition(Function *F,
+                                   std::vector<Backend *> backends) {
+  FunctionToBackendKindMapTy ret;
+  NodeToFunctionMap mapping;
+  std::unordered_map<Node *, BackendKind> nodeToBackendKind;
+
+  // For each node find a backend that supports it.
+  for (auto &N : F->getNodes()) {
+    for (auto &backend : backends) {
+      // Find the first backend that supports this node. The order of backends
+      // is important.
+      if (backend->shouldLower(&N) || backend->isOpSupported(N)) {
+        // Put this node into a partition for this backend.
+        nodeToBackendKind[&N] = backend->getBackendKind();
+        break;
+      }
+    }
+    assert(nodeToBackendKind.find(&N) != nodeToBackendKind.end() &&
+           "Node is not supported by any of the provided backends");
+  }
+
+  BFSLevel bfs = getBFSLevel(F);
+  size_t level = bfs.size();
+  int color = 0;
+  Function *newF;
+  newF = F->getParent()->createFunction(std::string(F->getName()) + "_part" +
+                                        std::to_string(++color));
+  BackendKind backendKind = nodeToBackendKind[bfs[level - 1][0]];
+  mapping.createPartition(newF, backendKind);
+  ret[newF] = backendKind;
+  for (int i = level - 1; i >= 0; i--) {
+    for (size_t j = 0, e = bfs[i].size(); j < e; j++) {
+      Node *N = bfs[i][j];
+      auto bk = nodeToBackendKind[N];
+      if (bk != backendKind) {
+        backendKind = bk;
+        newF = F->getParent()->createFunction(
+            std::string(F->getName()) + "_part" + std::to_string(++color));
+        mapping.createPartition(newF, backendKind);
+        ret[newF] = backendKind;
+      }
+      mapping.add(N, newF);
+    }
+  }
+
+  // Here we just need to split the function without generating DAG.
```

Comment:
Could it be that this remaining piece of code until the end of the function is very similar to another partitioning function? If so, could it be moved into a helper function and used from both?

Reply:
Do you plan to address this part in this PR?

Reply:
Sorry missed this part... Improved the code in this version. Thanks!